Add logic to convert declarativeNetRequest rules to WebKit content blocker rules

b-weinstein opened the pull request:
#### f3d8deeacd7258224aacd25276f66e9d1d8e242a
<pre>
Add logic to convert declarativeNetRequest rules to WebKit content blocker rules
<a href="https://bugs.webkit.org/show_bug.cgi?id=264982">https://bugs.webkit.org/show_bug.cgi?id=264982</a>
<a href="https://rdar.apple.com/118528800">rdar://118528800</a>

Reviewed by Timothy Hatcher.

This PR also gets rid of all of the `__kindof`s in the CocoaHelpers template functions. It is causing compile errors
on older OSes.

* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestRule.h: Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestRule.mm: Added.
(-[_WKWebExtensionDeclarativeNetRequestRule _validateHeaderInfoDictionary:]): Validates the header info dictionary.
(isHeaderNameValid): Checks the name of the header against a list of known headers.
(isArrayOfDomainsValid): Makes sure there are domains and they can all be converted to ASCII.
(isArrayOfExcludedDomainsValid): Makes sure all domains can be converted to ASCII.
(-[_WKWebExtensionDeclarativeNetRequestRule removeInvalidResourceTypesForKey:]): Remove any resource types WebKit can&apos;t handle.
(-[_WKWebExtensionDeclarativeNetRequestRule ruleInWebKitFormat]): Convert the rule to WebKit format.
(-[_WKWebExtensionDeclarativeNetRequestRule _convertedRulesForWebKitActionType:chromeActionType:]): Convert action type from Chrome
to WebKit format.
(-[_WKWebExtensionDeclarativeNetRequestRule _webKitRuleWithWebKitActionType:chromeActionType:chromeResourceTypes:]): Create a content blocking
rule with the given parameters.
(-[_WKWebExtensionDeclarativeNetRequestRule _chromeResourceTypeToWebKitLoadContext]): Convert a resource type to a load context.
(-[_WKWebExtensionDeclarativeNetRequestRule _chromeDomainTypeToWebKitDomainType]): Convert a Chrome domain type to a WebKit domain type.
(-[_WKWebExtensionDeclarativeNetRequestRule _chromeResourceTypeToWebKitResourceType]): Convert resource types.
(-[_WKWebExtensionDeclarativeNetRequestRule _resourcesToTargetWhenNoneAreSpecifiedInRule]):
(-[_WKWebExtensionDeclarativeNetRequestRule _allChromeResourceTypes]):
(-[_WKWebExtensionDeclarativeNetRequestRule _convertedResourceTypesForChromeResourceTypes:]):
(-[_WKWebExtensionDeclarativeNetRequestRule _regexURLFilterForChromeURLFilter:]):
(-[_WKWebExtensionDeclarativeNetRequestRule compare:]): Compare two rules.
(priorityForRuleType):
(-[_WKWebExtensionDeclarativeNetRequestRule description]): Print out information about the rule.
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestTranslator.h: Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestTranslator.mm: Added.
(+[_WKWebExtensionDeclarativeNetRequestTranslator translateRules:errorStrings:]): Translate the rules.
(+[_WKWebExtensionDeclarativeNetRequestTranslator jsonObjectsFromData:errorStrings:]): Perform JSON parsing.
(+[_WKWebExtensionDeclarativeNetRequestTranslator jsonStringForJSONObject:]): Ditto.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj: Add the new files.

Canonical link: <a href="https://commits.webkit.org/270861@main">https://commits.webkit.org/270861@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/735a4bd038f13e17ddf8a25853bbf9ca001c34da

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26663 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5275 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27901 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28875 "Built successfully") | [❌ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24385 "Failed to checkout and rebase branch from PR 20630") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7093 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2673 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/24296 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26924 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4095 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22885 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3600 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/3639 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29360 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24309 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24276 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29905 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/3676 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/1870 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27804 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5124 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6397 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/4149 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4027 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestTranslator.mm:
```diff
@@ -0,0 +1,108 @@
+    ASSERT([NSJSONSerialization isValidJSONObject:jsonObject]);
+    NSData *jsonAsData = [NSJSONSerialization dataWithJSONObject:jsonObject options:0 error:nil];
+    return [[NSString alloc] initWithData:jsonAsData encoding:NSUTF8StringEncoding];
+}
```
There is a CocoaHelpers function that does this already.

b-weinstein replied:
These aren't used yet - so I'll just get rid of them now and adopt the CocoaHelpers ones when adopting these classes.

xeenon commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -381,6 +381,24 @@ id parseJSON(API::Data& json, JSONOptionSet options, NSError **error)
+
+    return stringHasCaseInsensitivePrefix(urlString, @"http:") || stringHasCaseInsensitivePrefix(urlString, @"https:");
+
+}
```
We can use WTF URL methods for this check, it can convert from NSURL easily with a cast.

b-weinstein replied:
I got rid of this helper method and used:

`if (!URL(url).protocolIsInHTTPFamily()) {`

at the callsite.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestTranslator.mm:
```diff
@@ -0,0 +1,108 @@
+    NSMutableArray<NSString *> *errors = [NSMutableArray array];
+    for (NSData *jsonData in jsonDataArray) {
+        NSError *error;
+        NSArray<NSDictionary *> *json = [NSJSONSerialization JSONObjectWithData:jsonData options:0 error:&error];
```
There is also a CocoaHelper for this.

b-weinstein replied:
Adopted `parseJSON`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestTranslator.mm:
```diff
@@ -0,0 +1,108 @@
+#import "_WKWebExtensionDeclarativeNetRequestTranslator.h"
+
+#import "_WKWebExtensionDeclarativeNetRequestRule.h"
+#import <wtf/Assertions.h>
```
Is this needed? I thought config.h included it.

b-weinstein replied:
I was able to get rid of this.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestTranslator.mm:
```diff
@@ -0,0 +1,108 @@
+
+#import "config.h"
+#if ENABLE(WK_WEB_EXTENSIONS)
+#import "_WKWebExtensionDeclarativeNetRequestTranslator.h"
```
Add new lines around #if.

b-weinstein replied:
Without the newlines I was getting a style error. I'll try again though.

b-weinstein replied:
ERROR: Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestTranslator.mm:30:  You should not add a blank line before implementation file's own header.  [build/include_order] [4]

xeenon replied:
Oh, move the second import out next to config.h, the rest should be inside the #if

b-weinstein committed:
Add logic to convert declarativeNetRequest rules to WebKit content blocker rules
https://bugs.webkit.org/show_bug.cgi?id=264982
rdar://118528800

Reviewed by Timothy Hatcher.

This PR also gets rid of all of the `__kindof`s in the CocoaHelpers template functions. It is causing compile errors
on older OSes.

* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestRule.h: Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestRule.mm: Added.
(-[_WKWebExtensionDeclarativeNetRequestRule _validateHeaderInfoDictionary:]): Validates the header info dictionary.
(isHeaderNameValid): Checks the name of the header against a list of known headers.
(isArrayOfDomainsValid): Makes sure there are domains and they can all be converted to ASCII.
(isArrayOfExcludedDomainsValid): Makes sure all domains can be converted to ASCII.
(-[_WKWebExtensionDeclarativeNetRequestRule removeInvalidResourceTypesForKey:]): Remove any resource types WebKit can't handle.
(-[_WKWebExtensionDeclarativeNetRequestRule ruleInWebKitFormat]): Convert the rule to WebKit format.
(-[_WKWebExtensionDeclarativeNetRequestRule _convertedRulesForWebKitActionType:chromeActionType:]): Convert action type from Chrome
to WebKit format.
(-[_WKWebExtensionDeclarativeNetRequestRule _webKitRuleWithWebKitActionType:chromeActionType:chromeResourceTypes:]): Create a content blocking
rule with the given parameters.
(-[_WKWebExtensionDeclarativeNetRequestRule _chromeResourceTypeToWebKitLoadContext]): Convert a resource type to a load context.
(-[_WKWebExtensionDeclarativeNetRequestRule _chromeDomainTypeToWebKitDomainType]): Convert a Chrome domain type to a WebKit domain type.
(-[_WKWebExtensionDeclarativeNetRequestRule _chromeResourceTypeToWebKitResourceType]): Convert resource types.
(-[_WKWebExtensionDeclarativeNetRequestRule _resourcesToTargetWhenNoneAreSpecifiedInRule]):
(-[_WKWebExtensionDeclarativeNetRequestRule _allChromeResourceTypes]):
(-[_WKWebExtensionDeclarativeNetRequestRule _convertedResourceTypesForChromeResourceTypes:]):
(-[_WKWebExtensionDeclarativeNetRequestRule _regexURLFilterForChromeURLFilter:]):
(-[_WKWebExtensionDeclarativeNetRequestRule compare:]): Compare two rules.
(priorityForRuleType):
(-[_WKWebExtensionDeclarativeNetRequestRule description]): Print out information about the rule.
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestTranslator.h: Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestTranslator.mm: Added.
(+[_WKWebExtensionDeclarativeNetRequestTranslator translateRules:errorStrings:]): Translate the rules.
(+[_WKWebExtensionDeclarativeNetRequestTranslator jsonObjectsFromData:errorStrings:]): Perform JSON parsing.
(+[_WKWebExtensionDeclarativeNetRequestTranslator jsonStringForJSONObject:]): Ditto.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj: Add the new files.

Canonical link: https://commits.webkit.org/270861@main
